Expand out offline app features for smoother cache clearing and version tracking

anthony-zhou opened the pull request:
Fixes #1364

So far, I've displayed the NPM version number at the bottom of the page by parsing it from the package.json file. I'm working on also adding the version number to the top right and detecting when a new version has successfully been fetched.

* [x] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `npm test`
* [x] code is in uniquely-named feature branch and has no merge conflicts
* [x] PR is descriptively titled
* [x] Insert-step functionality is working correct as expected.

## Screenshots
Here is a gif of the version number text, in the top right and at the bottom of the page:
![Image Sequencer Version Number Screenshot](https://user-images.githubusercontent.com/37421108/71702753-c57efe80-2d96-11ea-998f-3b4342287ce5.gif)

Here is a screenshot of the prompt to update to the latest version (in the top left):
![image](https://user-images.githubusercontent.com/37421108/71702836-30303a00-2d97-11ea-8199-08872d04a1c5.png)

anthony-zhou commented:
@HarshKhandeparkar I'm working on the feature you mentioned, but as part of this I'd like to use the Node module [semver](https://www.npmjs.com/package/semver). For some reason the traditional `const semver = require('semver')` is throwing a Javascript error of 
```
image-sequencer.js:1 Uncaught Error: Cannot find module './functions/lt'
    at o (image-sequencer.js:1)
    at o (image-sequencer-ui.js:1)
    at o (image-sequencer-ui.js:1)
    at image-sequencer-ui.js:1
    at lazyRequire (image-sequencer-ui.js:47990)
    at Object.get [as lt] (image-sequencer-ui.js:47997)
    at XMLHttpRequest.request.onreadystatechange (image-sequencer-ui.js:48)```

Would you happen to know what is causing this error?

keshav234156 commented:
@anthony-zhou Do you know Service Workers? It is basically a script that browsers in run in the background that enable to control network request and cache those requests. Each service Worker has a lifecycle namely registration, Installation and Activation and installation is triggered only if you visit the application for the first time or there is some difference in the new service worker. So we can add event listeners to the service worker. Then we can check whether service worker was installed or not and do all the required stuff.
Thanks

HarshKhandeparkar commented on examples/demo.js:
```diff
@@ -28,6 +32,33 @@ window.onload = function () {
+        }
+      }
+    }
+    request.open("GET", "https://api.github.com/repos/publiclab/image-sequencer/contents/package.json", true);
```
This is cool. So you want to clear the cache(via demo.js) if a new version is available?

anthony-zhou replied:
Yeah, but I'll probably end up doing this through the service worker.

HarshKhandeparkar replied:
This is not a bad way either. Is there a downside to this?

anthony-zhou commented:
I made some more changes:
- Moved fetching of local and latest version number to versionManagement.js.
- Added a popup in index.html that is displayed when a new service worker is loaded.
- This works by loading a new service worker when `sw.js` is changed.

Currently, the popup I created is displayed whenever the `sw.js` file is changed, which means that when a new version of Image Sequencer is deployed we would also need to change the version number in the first line of `sw.js`. Then we could compare the local version number to the online version number and get the new version if needed. 

I'm thinking it might be inconvenient to update sw.js each time a new version is deployed. Is there any way we could do this automatically? @HarshKhandeparkar @keshav234156

HarshKhandeparkar commented:
I don't think doing it automatically will be easy. We can't browserify it. We will either have to make a script which changes the specific line in the code or let `sw.js` fetch changes from the package.json file.

anthony-zhou commented:
I just pushed a commit that adds a faint version number in fixed position in the top right corner.

So now, I believe the app should fetch the newest version whenever a new version is deployed, as long as `sw.js` is changed when we deploy a new version. So the changes technically work now--the question is whether we need to implement automatic updates to `sw.js`. This would make deploying a new version easier but would require running an additional script each time changes are made to `package.json`.

In order to update `sw.js` automatically, I think the best option may be to run a script that changes the `sw.js` file whenever `package.json` is changed. However, this would probably require editing some of the configuration files here on GitHub. Does it make sense to add a script in GitHub that does this (presumably it would have to be run whenever changes are made to the repository to check if changes were made to `package.json`)?

anthony-zhou commented:
@keshav234156 @HarshKhandeparkar I ended up creating a grunt task to update `sw.js` whenever `grunt build` is run. I used the dependency `grunt-text-replace` to replace the text for the version number. Now everything should be working as expected. I would appreciate if you could review this PR again. Thanks!

HarshKhandeparkar commented on Gruntfile.js:
```diff
@@ -78,7 +90,7 @@ module.exports = function(grunt) {
   /* Default (development): Watch files and build on change. */
   grunt.registerTask('default', ['watch']);
-  grunt.registerTask('build', ['browserify:core', 'browserify:ui', 'uglify:core', 'uglify:ui']);
+  grunt.registerTask('build', ['browserify:core', 'browserify:ui', 'replace:version', 'uglify:core', 'uglify:ui']);
```
Can we do the replace version in the `production` and `serve` tasks too?

anthony-zhou replied:
Done.

HarshKhandeparkar commented on examples/lib/versionManagement.js:
```diff
@@ -0,0 +1,86 @@
+      callback(latestVersionNumber);
+    }
+  }
+  request.open("GET", "https://api.github.com/repos/publiclab/image-sequencer/contents/package.json", true);
```
This will GET the latest package JSON. It may not be the one online. We need a better way of GET-ing the version number.

anthony-zhou replied:
By online, do you mean at sequencer.publiclab.org? If so, it looks like we could get the version number from the `stable` branch here on GitHub. 

Whether we do this depends on whether offline users are mainly working off the `stable` branch or the `main` branch. The current implementation of this function is getting the version number from the `main` branch, which may be what people are using offline anyways.

HarshKhandeparkar replied:
For beta.sequencer.publiclab.org, we'll have to get the version from that repo. Beta is on a separate repo btw. For the stable one, it will bw different. Anyway, it will be a mess. Can't we get it from `current_location/package.json` or something? Make it dynamic instead of hard coded?

HarshKhandeparkar commented on examples/lib/versionManagement.js:
```diff
@@ -0,0 +1,86 @@
+  return v1parts.length != v2parts.length;
+}
+
+function compare(part1, part2) {
```
The parts of a version number are called `major.minor.patch`. Could you please use that terminology here too? To make it more readable.

HarshKhandeparkar commented on package.json:
```diff
@@ -70,6 +70,7 @@
     "readline-sync": "^1.4.7",
     "save-pixels": "~2.3.4",
     "selectize": "^0.12.6",
+    "semver": "^7.1.1",
```
Is semver used now?

anthony-zhou replied:
Good point. I have now uninstalled it.

HarshKhandeparkar commented on examples/lib/versionManagement.js:
```diff
@@ -0,0 +1,86 @@
+// Get the current version number from package.json on GitHub.
```
Please add a few more comments to this file to make it easily understandable.

anthony-zhou commented:
@HarshKhandeparkar Thanks for all the feedback! I removed the versionCompare() function in versionManagement.js, which wasn't being used anyways and was responsible for most of the messy code in that file, so hopefully that addresses several of your comments. Please see my individual replies for more details.

SidharthBansal commented:
I am impressed from your collaboration. Collaboration is the essence of the Public Lab community.
Harsh please get points for https://codein.withgoogle.com/dashboard/tasks/6206382643085312/.
I made the instance count as 2. I hope @anthony-zhou  both are ok with it. 

In case you think you are doing more work and should get points for it, please ping me. We need to be fair to all.

SidharthBansal commented:
@HarshKhandeparkar you will get 5 points for this PR review task as you are helping them a lot. Thanks both of you.
cc\ @jywarren

HarshKhandeparkar commented:
Thanks. I will drop another review tomorrow. I am a bit busy today. Sorry
about that.

On Wed, 1 Jan, 2020, 9:54 PM Sidharth Bansal, <notifications@github.com>
wrote:

> @HarshKhandeparkar <https://github.com/HarshKhandeparkar> you will get 5
> points for this PR review task as you are helping them a lot. Thanks both
> of you.
> cc\ @jywarren <https://github.com/jywarren>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/image-sequencer/pull/1384?email_source=notifications&email_token=AIJI5HYG6ZBW4JDOMTRSGPLQ3S7TJA5CNFSM4J6JWM72YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEH5ID5Q#issuecomment-570065398>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIJI5H6TMRLAZ3EWRQGHFCDQ3S7TJANCNFSM4J6JWM7Q>
> .
>

anthony-zhou commented:
@HarshKhandeparkar I changed the URL in the getLatestVersion function to get the package.json file from the local package.json file's attribute `homepage`, which is sequencer.publiclab.org for now. This could cause issues if the domain is changed, but hopefully that's not likely.

I was tempted to use something like `window.location.hostname` or `window.location.href` but that wouldn't get the online version of package.json if Image Sequencer is running locally.

keshav234156 commented:
@jywarren please review the changes

jywarren commented:
Wow this is really ambitious! What reference did you ultimately choose for `homepage` vs. the `CNAME` record? I think the CNAME file is only in the gh-pages branch maybe? Not sure!

As this does not have a test, would you please upload a GIF demonstrating it working? 

If you'd like perhaps we can make a new task for writing a test for this? That would be great and would help ensure this functions as intended as the codebase evolves.

anthony-zhou commented:
@jywarren I've uploaded a GIF and screenshot in the PR description showing the new features. How do they look?

As for making a test, I'd be happy to do that as another task. What exactly would that involve? 

Also, I think the `homepage` reference works better for now, although it seems like neither that nor CNAME would get the beta URL as @HarshKhandeparkar mentioned. Either way, we will have the latest stable version number. The version number displayed on the page is the actual version number being used (the "local" version number), so this shouldn't be an issue.

Another note: `window.location` will just return localhost if it's being run offline, so it won't get the latest version number from online.

HarshKhandeparkar commented:
We don't need the online version number. We want the local version number. We want the version number that the user is accessing. Locally, the user accesses a server on the local PC. That's the version we want to install.

anthony-zhou commented:
@HarshKhandeparkar The way I have it set up now, it is installing from a local version. Right now, the function for getting the latest version (`getLatestVersionNumber`) isn't really being used for anything. It is `getLocalVersionNumber` that is being used to find the latest local version to install.

HarshKhandeparkar commented:
Oh. What will it be used for in the future?

anthony-zhou commented:
Not sure. It could be used to let users know when they should pull a new version from online, but it seems that many npm packages and related apps don't do that. Do you think we should keep this function?

anthony-zhou commented:
@jywarren @HarshKhandeparkar I understand that it's a busy time but I'd really appreciate if I could get reviewed for this pull request ASAP so we can move towards publishing this. I'm really excited to work on other tasks for GCI, including a test for the current feature, but I can't move on until this task is approved. Meanwhile, I have been working on FTOs and other supportive tasks. Yet I feel that little progress has been made in the past week... if there is anything I should work on to improve this PR, please let me know. Thanks!

anthony-zhou commented:
Also, I have kept `getLatestVersionNumber` in `versionManagement.js` as it can be used as a debug statement.

keshav234156 commented:
@jywarren CAn you please review this ASAP

jywarren commented:
OK, thanks for your patience! I'll review this next!

jywarren commented:
OK, my understanding of `getLatestVersionNumber()` was that we were going to use it to establish whether https://sequencer.publiclab.org (or beta, but let's put that aside for now for simplicity) has a newer version available than what's been cached. So it should compare:

1. the npm version of the cached app TO
2. the npm version of the app available at sequencer.publiclab.org

Is that right? OK, now I'll look through the latest files to see how this connects together. Thanks!

jywarren commented on examples/lib/cache.js:
```diff
@@ -1,7 +1,32 @@
   if ('serviceWorker' in navigator) {
     navigator.serviceWorker.register('sw.js', { scope: '/examples/' })
       .then(function(registration) {
+        registration.addEventListener('updatefound', () => {
```
So, here we can detect if there's a new version, but we don't actually tie this to `getLatestVersionNumber()`, as far as I can read it? How might we do that - could we say `Click to install version X` here? Or is `getLatestVersionNumber()` not reading from the same source as `updatefound` here?

jywarren replied:
We don't have to change things; i like how this looks. But let's leave enough inline comments to make the answer to this question really clear! Esp. because this is complicated stuff, so future coders should be able to follow what you've done. Thanks!

jywarren commented:
And thanks for your patience! Next time, perhaps we can work in smaller chunks so that it can be merged in a modular fashion, and we won't get so slowed down, as it seems this PR has several portions which are already complete and could be merged on their own!

Thanks, all!

anthony-zhou commented:
@jywarren Thanks for the feedback. I added some explanatory comments to `versionManagement.js` , `cache.js` and `sw.js` in the areas I worked on. Is there anything else to add?

Also, good suggestion about working in a modular fashion... I'll be sure to do that next time.

jywarren commented on examples/lib/versionManagement.js:
```diff
@@ -1,14 +1,28 @@
+/**
+ * Functions for getting version information.
+ * Note: these functions are not used by the service worker to check for updates;
+ * the service worker updates whenever sw.js has changed.
```
How about one more bit here... like:

> ...and sw.js is changed when `grunt ____` is run

? Thank you!

anthony-zhou replied:
@jywarren Sounds good. I have added that change.

anthony-zhou replied:
@jywarren Sounds good. I have added that change.

jywarren commented:
Oh no! Tests didn't pass this time. Was there a typo perhaps?

HarshKhandeparkar commented on examples/index.html:
```diff
@@ -38,7 +38,7 @@
   <script src="../node_modules/jspdf/dist/jspdf.min.js" type="text/javascript" ></script>
 
-  <script src="lib/scrollToTop.js"></script>
+  <!-- <script src="lib/scrollToTop.js"></script> -->
```
Does this thing not exist?

anthony-zhou replied:
Seems like it. Not sure what it does or where it might be. I commented out this link to clear up the error on the console.

HarshKhandeparkar commented on examples/lib/cache.js:
```diff
@@ -1,7 +1,39 @@
+    $('#update-prompt-modal').addClass('show');
+  }
+
+  // When a new service worker has been loaded, the button in the update prompt
```
Please use multi-line comments instead of multiple inline comments. That will make it easier to read. Also, please keep words like service and worker on the same line.

anthony-zhou replied:
Got it, thanks!

HarshKhandeparkar commented on examples/lib/versionManagement.js:
```diff
@@ -0,0 +1,40 @@
+ */
+function getLatestVersionNumber(callback) {
+  // Get the homepage reference from the local package.json.
+  var homepage = require('../../package.json').homepage;
```
Awesome!

HarshKhandeparkar commented on examples/lib/versionManagement.js:
```diff
@@ -0,0 +1,40 @@
+      var response = JSON.parse(this.responseText);
+      var latestVersionNumber = response.version;
+
+      // Do something with the version number using a callback function.
```
Is this callback optional?

anthony-zhou replied:
It's needed to use the version number fetched, but it's probably still good practice to check if it is defined using `if (callback)`.

HarshKhandeparkar commented on examples/lib/versionManagement.js:
```diff
@@ -0,0 +1,40 @@
+      var latestVersionNumber = response.version;
+
+      // Do something with the version number using a callback function.
+      callback(latestVersionNumber);
```
If it is optional then please add an
```
if (callback)
```

HarshKhandeparkar commented on examples/lib/versionManagement.js:
```diff
@@ -0,0 +1,40 @@
+
+// Get the version number from the local package.json file.
+function getLocalVersionNumber() {
+  return require('../../package.json').version;
```
Why does the require statement appear twice?

HarshKhandeparkar replied:
Also, I think adding all the require statements at the start of the file is a good practice.

anthony-zhou replied:
I moved the require statement to the start of the file, so it's now called only once.

HarshKhandeparkar commented on examples/sw.js:
```diff
@@ -1,5 +1,4 @@
-const staticCacheName = 'image-sequencer-static-v3';
-
+const staticCacheName = 'image-sequencer-static-v3.5.1';
```
Cool.

HarshKhandeparkar commented:
I think travis is broken. It had broken before. It could fix itself like before. Perhaps the servers are down or there are too many requests.

jywarren commented:
It's possible Travis is down... but looks ok here; https://www.traviscistatus.com/

jywarren commented:
Oh it passed now! cool. But it's out of sync. I'll sync it.

jywarren commented:
Just waiting on some answers to @HarshKhandeparkar's questions, then ready to merge this!

anthony-zhou commented:
@jywarren @HarshKhandeparkar Thanks for reviewing. I have implemented all the changes you mentioned.

jywarren commented:
Fantastic. Waiting for this to pass before merging. Thanks a ton, @anthony-zhou !!!

jywarren commented:
Whew!!! Great!!!!! Thanks a ton!
